tasks/install/containers/container.yml

```diff
@@ -15,6 +19,24 @@
     mode: 0644
   register: has_new_config_changed
 
+  - name: "Copying eas_produce.json"
+  copy:
+    src: "{{playbook_dir}}/eas.d/eas_produce.json"
```

Comment:
Warum `{{ playbook_dir }}`? Dadurch entsteht einer harte Kopplung zwischen Rolle und Playbook, die eigentlich extra entkoppelt sind. Besser waere vermutlich die Dateien hier in die Rolle nach `files/` zu legen. Was meinst du?

Reply:
Ich habe das so gestaltet, damit die Konfigurationsdateien von einem Kunden genau an einem Ort liegen. Versionen der easydb5 können leider auch für jeden Kunden unterschiedlich sein.

Wenn du das aber für eine schlechte Idee hälst, kann ich das gerne umbauen.

Reply:
Hm... kommt drauf an... sind die Konfigurationsdateien denn so unterschiedlich, dass die sich nicht mit Variablen parametrisieren lassen?

Reply:
Hier mal ein kleiner Ausschnitt einer eas_rights_management.yml:

```yml
eas:
  rights_management:
    image:
      versions:
        - version: huge
          size_print: Pressebild
          size_limit: 1600
          export: true
          rightsmanagement: true
          group: preview
          zoomable: true
        - version: jpg1088
          size_print: 1088px
          size_limit: 1088
          export: true
          group: preview
          rightsmanagement: true
          zoomable: true
          watermark: true
    video:
      versions:
        - version: 720p
          size_print: 720p
          size_limit: 1280
          export: true
          group: preview
          rightsmanagement: true
        - version: 360p
          size_print: 360p
          size_limit: 640
          export: true
          group: preview
          rightsmanagement: true
    audio:
      versions:
        - version: aac
          size_print: aac
          export: true
          rightsmanagement: true
        - version: preview
          size_print: 640px
          rightsmanagement: true
    office:
      versions:
        - version: pdf
          size_print: pdf
          export: true
          rightsmanagement: true
        - version: huge
          size_print: 1600px
          rightsmanagement: true
    archive:
      versions:
        - version: directory
          size_print: directory
          group: huge
          rightsmanagement: true
        - version: small
          size_print: 200px
          size_limit: 200
          rightsmanagement: false
          group: thumbnail
          standard: true
    unknown:
      versions: []
    vector2d:
      versions: []
    vector3d:
      versions: []
```

und ein Ausschnitt der eas_produce.json
```json
{
    "__all": {
        "__all": {
            "small": {
                "target_size": "1000x250",
                "target_format": "png",
                "target_alpha": "on",
                "target_no_fallback": "1",
                "priority": "12"
            }
        }
    },

    "image": {
        "__all": {
            "huge": {
                "target_size": "1600x1600",
                "target_format": "jpg",
                "target_interlace": "1",
                "target_no_enlarge": "1"
            },
            "jpg1088": {
                "target_size": "1088x1088",
                "target_format": "jpg",
                "target_interlace": "1",
                "target_no_enlarge": "1"
            }
    },
    "video": {
        "__all": {
            "preview": {
                "target_size": "960x640",
                "target_format": "jpg",
                "target_interlace": "1",
                "target_no_enlarge": "1"
            },
            "360p": {
                "target_height": "360",
                "target_format": "mp4",
                "target_no_enlarge": "1",
                "target_audio_bitrate": "160k",
                "target_video_bitrate": "840k"
            }
    },
    "audio": {
        "__all": {
            "preview": {
                "target_size": "640x640",
                "target_format": "jpg",
                "target_interlace": "1",
                "target_no_fallback": "1"
            },
            "aac": {
                "target_format": "aac",
                "target_audio_bitrate": "160k"
            }
        }
    },
    "office": {
        "__all": {
            "__pages": {
                "__source": {
                    "target_format": "png",
                    "target_size": "1000x1000"
                },
                "small": {
                    "target_size": "800x800",
                    "target_format": "png"
                }
            },
            "pdf": {
                "target_format": "pdf"
            },
            "preview": {
                "target_size": "960x960",
                "target_format": "jpg",
                "target_interlace": "1",
                "target_no_enlarge": "1"
            },
            "huge": {
                "target_size": "1600x1600",
                "target_format": "jpg",
                "target_no_enlarge": "1"
            }
        },
        "pdf": {
            "pdf": false
        }
    },

    "archive": {
        "__all": {
            "small": false
        },
        "webdvd.zip": {
            "directory": {
                "target_format": "directory"
            },
            "small": {
                "target_size": "200x200",
                "target_format": "png",
                "target_alpha": "on",
                "target_no_fallback": "1",
                "source_version": "directory",
                "priority": "12"
            }
        }
    },

    "unknown": {
    }
}

Reply:
Die ist jetzt schon ziemlich gekürzt.

Reply:
> Hm... kommt drauf an... sind die Konfigurationsdateien denn so unterschiedlich, dass die sich nicht mit Variablen parametrisieren lassen?

Ja, parametrisieren ist hier sehr aufwändig(Liste von yaml-Abschnitten, jeder Davon mit vielen Parametern, die auch ganz fehlen können. Die Abschnitte können auch mal in einen anderen Teil der Hierarchie gehören als sonst. Manchmal sind sie zusätzlich, manchmal anstatt.
Besser die Datei als Ganzes kopieren. 

Kann mir außerdem gut vorstellen dass so ein Fall noch öfter auftaucht, also: Datei-ist-ganz-anders-pro-Kunde, kopier-es-einfach-rüber. Also macht es Sinn hier einen Weg zu kennen.

In der role-apache2-vhost dachte ich schon wir hätten das auch, für die Dateien in sites-enabled/easydb5/include-before und -after. Oder legen wir hier die Dateien immer per Hand rein?

Reply:
> Warum `{{ playbook_dir }}`? Dadurch entsteht einer harte Kopplung zwischen Rolle und Playbook, die eigentlich extra entkoppelt sind. Besser waere vermutlich die Dateien hier in die Rolle nach `files/` zu legen. Was meinst du?

Welches files/ meinst Du? Bitte dazuschreiben. In der Rolle will ich nicht plötzlich pro-Kunden-Dateien ablegen.  DAS wäre eine Kopplung die ich nicht gut finde.
Und wo im playbook? files/eas_rights_management.json/KUNDENKUERZEL? Oder KUNDENKUERZEL/files/eas_rights_management.json ?

Reply:
Gut, ich ürde evtl. trotzdem noch dahingehend umformulieren, dass die Rolle auch dann erfolgreich durchläuft, wenn die Datei im Playbook nicht vorhanden ist. Also entweder ein `when:` was nur triggert wenn die Datei existiert oder wenn ein boolean Flag (default: False) auf True gesetzt ist. Spricht da was dagegen?

Reply:
Oder können wir den relativen Pfad im playbook setzen statt in der Rolle? Pseudo-config in playbook/KUNDENKUERZEL/vars/easydb5.yml:

```
eas_produce_file: eas_rights_management.yml
```
... meine Hoffnung wäre dann dass der relative Pfad auf `playbook/KUNDENKUERZEL/vars/eas_rights_management.yml` zeigt. Geht das?

in role/tasks/...:
```
copy:
  src: {{{eas_produce_file}}
```

Reply:
@Leon: Bis wir eine bessere Lösung haben können wir mit {{ playbook_dir }} arbeiten.

Reply:
Ich meine das `files/` in der Rolle, so wie ich es dazu geschrieben habe. Über diesen Punkt waren wir in dieser Diskussion aber schon hinaus, denn da die Datei pro Kunde individuell ist, gehört sie nicht in die Rolle - der Ansatz von Leon ist also schon gut, dass die Datei von der Rolle aus dem `{{ playbook_dir }}` gezogen wird, weil da die kundenspezifischen Dinge im Playbook festgehalten sind. Ich nehme aber trotzdem an, dass die Rolle auch dann durchlaufen soll, wenn ein solches File im Playbook nicht existiert. Daher rührt mein Vorschlag da noch ein `when:´ dazu zu schreiben, welches überprüft, ob die Datei denn überhaupt im Playbook existiert, bevor es versucht eine nicht-existente Datei hochzuladen.

Reply:
Das ist auch enthalten:
```yml
- name: "Copying eas_produce.json"
  copy:
    src: "{{playbook_dir}}/eas.d/eas_produce.json"
    dest: "{{ easydb_basedir}}/config/eas_produce.json"
    owner: root
    group: root
    mode: 0644
  when: easydb_eas_custom_version_enabled == True or easydb_eas_watermark_enabled == True
  
- name: "Copying eas_rights_management.json"
  copy:
    src: "{{playbook_dir}}/eas.d/eas_rights_management.yml"
    dest: "{{ easydb_basedir}}/config/eas_rights_management.yml"
    owner: root
    group: root
    mode: 0644
  when: easydb_eas_custom_version_enabled == True or easydb_eas_watermark_enabled == True
```
In der ```role-easydb5/tasks/install/containers/eas.yml```